### PR TITLE
Consider MPP FDW LIMIT pushdown when both offset and limit clause are specified

### PR DESCRIPTION
**Pull request by jingwen-yang-yjw**
### What this PR does
Before this PR, GPDB won't push OFFSET/LIMIT in the remote servers for MPP FDW when both offset and limit clause are specified.
Because it's **unsafe** if we don't adjust `Remote SQL`.

In this PR, we consider MPP FDW OFFSET/LIMIT pushdown when both OFFSET and LIMIT clause are specified.
We will adjust `Remote SQL` when `mpp_execute = 'all segments'` for query including OFFSET/LIMIT clause.

### Implementation Details
There are 3 cases for query including OFFSET/LIMIT clause.

**Case 1: only LIMIT is specified**
Two-phase LIMIT plan is enough and there is NO need to adjust `Remote SQL`.
This case had been supported. (Look at PR https://github.com/greenplum-db/gpdb/pull/16177 for more information.)
   
**Case 2: both LIMIT and OFFSET are specified**
We also need to adjust `Remote SQL` to let GPDB fetch tuples from 0 to limitOffset + limitCount from remote servers.
```
-- Simple query with OFFSET and LIMIT clause together is pushed down.
EXPLAIN (VERBOSE, COSTS OFF)
SELECT c1, c2 FROM mpp_ft2 order by c1 offset 2 limit 3;
                                          QUERY PLAN                                           
-----------------------------------------------------------------------------------------------
 Limit
   Output: c1, c2
   ->  Gather Motion 2:1  (slice1; segments: 2)
         Output: c1, c2
         Merge Key: c1
         ->  Foreign Scan on public.mpp_ft2
               Output: c1, c2
               Remote SQL: SELECT c1, c2 FROM "MPP_S 1"."T 2" ORDER BY c1 ASC NULLS LAST LIMIT (2::bigint + 3::bigint)
 Optimizer: Postgres-based planner
 ```
   

**Case 3: only OFFSET is specified**
   In this case, we also need to adjust `Remote SQL` to let GPDB fetch all tuples of foreign table from remote servers.
```
-- Simple query with only OFFSET clause is NOT pushed down.
EXPLAIN (VERBOSE, COSTS OFF)
SELECT c1, c2 FROM mpp_ft2 order by c1 offset 998;
                                       QUERY PLAN                                        
-----------------------------------------------------------------------------------------
 Limit
   Output: c1, c2
   ->  Gather Motion 2:1  (slice1; segments: 2)
         Output: c1, c2
         Merge Key: c1
         ->  Foreign Scan on public.mpp_ft2
               Output: c1, c2
               Remote SQL: SELECT c1, c2 FROM "MPP_S 1"."T 2" ORDER BY c1 ASC NULLS LAST
 Optimizer: Postgres-based planner
 ```

**Pipeline**: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb-dev-improve_mpp_offset_limit_pushdown-rocky8?group=all

### CODE REVIEW

**Comment by avamingli** [comment]
Hi, is there any links to tell about the grammar LIMIT(x + y) ?

> Remote SQL: SELECT c1, c2 FROM "MPP_S 1"."T 2" ORDER BY c1 ASC NULLS LAST LIMIT (2::bigint + 3::bigint)

**Comment by avamingli** [comment]
> Hi, is there any links to tell about the grammar LIMIT(x + y) ?
> 
> > Remote SQL: SELECT c1, c2 FROM "MPP_S 1"."T 2" ORDER BY c1 ASC NULLS LAST LIMIT (2::bigint + 3::bigint)

I get your point, sum offset and limit here for final limit node with offset.

**Comment by jingwen-yang-yjw** [comment]
> Hi, is there any links to tell about the grammar LIMIT(x + y) ?
> 
> > Remote SQL: SELECT c1, c2 FROM "MPP_S 1"."T 2" ORDER BY c1 ASC NULLS LAST LIMIT (2::bigint + 3::bigint)

I don't find explicit LIMIT(x + y) syntax introduction in official documents.

But in https://www.postgresql.org/docs/12/sql-select.html#SQL-LIMIT
```
If the count expression evaluates to NULL, it is treated as LIMIT ALL, i.e., no limit. If start evaluates to NULL, it is treated the same as OFFSET 0.
```
_count_ can be an expression for `LIMIT count` clause.
And the type of _count_ is `a_expr` in gram.y.

**Review by zhaorui20005 — APPROVED** [review]
LGTM

File: contrib/postgres_fdw/expected/mpp_gp2pg_postgres_fdw.out [hunk 1]
```diff
@@ -913,19 +908,19 @@ ALTER FOREIGN TABLE mpp_ft2 OPTIONS(set use_remote_estimate 'false');
 -- Queries with LIMIT/OFFSET clauses
 -- ===================================================================
 -- Simple query with LIMIT clause is pushed down.
-EXPLAIN (VERBOSE, COSTS OFF)
+EXPLAIN VERBOSE
```
Comment (huiliang-liu) [inline contrib/postgres_fdw/expected/mpp_gp2pg_postgres_fdw.out:911]: why remove `COSTS OFF` for the explain?

Comment (jingwen-yang-yjw) [reply contrib/postgres_fdw/expected/mpp_gp2pg_postgres_fdw.out:911]: > why remove `COSTS OFF` for the explain?

I adjust cost estimation for query with LIMIT/OFFSET when mpp_execute = 'all segments' in commit https://github.com/greenplum-db/gpdb/pull/16919/commits/1d439670c5a9cdff2dfeff848bc7714f3f9153d7.
So I remove `COSTS OFF` to test if the row estimation is correct or not.

**Review by huiliang-liu — APPROVED** [review]
Thank you for your contribution!

**Comment by kainwen** [comment]
What if the limit clause contains volatile functions?

**Comment by jingwen-yang-yjw** [comment]
> What if the limit clause contains volatile functions?

Before adding LIMIT/OFFSET pushdown path, postgres_fdw will make a check for LIMIT/OFFSET expressions.
If LIMIT/OFFSET expressions contain volatile functions, postgres_fdw will consider it unsafe and won't generate LIMIT/OFFSET pushdown path.
https://github.com/greenplum-db/gpdb/blob/96d47d3ec827812e10525274769c8a328036f593/contrib/postgres_fdw/postgres_fdw.c#L6773-L6779
https://github.com/greenplum-db/gpdb/blob/96d47d3ec827812e10525274769c8a328036f593/contrib/postgres_fdw/deparse.c#L359-L367

**Review by kainwen — APPROVED** [review]

**Pull request merged**